sycl/doc/Assert.md

```diff
@@ -0,0 +1,426 @@
+
+The format of the assert message is unspecified, but it will always include the
+text of the failing expression, the values of the standard macros `__FILE__` and
+`__LINE__`, and the value of the standard variable `__func__`. If the failing
+assert comes from an `nd_range` `parallel_for` it will also include the global
+ID and the local ID of the failing work item.
+
+Implementation of this function is supplied by Native Device Compiler for
+safe approach or by DPCPP Compiler for fallback one.
+
+NB: Due to lack of support of online linking in Level-Zero, the application is
+linked against fallback implementation of `__devicelib_assert_fail`. Hence,
+Native Device Compilers should prefer their implementation instead of the one
+provided in incoming SPIR-V/LLVM IR binary.
+
+
+## Safe approach
+
+This is the preferred approach and implementations should use it when possible.
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+When detected, Low-level Runtime reports assert failure to DPCPP Runtime
+via events objects. To achieve this, information about assert failure should be
+propagated from device-side to SYCL Runtime. This should be performed via calls
+to `piEventGetInfo`. This Plugin Interface call "lowers" to `clGetEventInfo` for
+OpenCL backend and `zeEventQueryStatus` for Level-Zero backend.
+
+Refer to [OpenCL](extensions/Assert/opencl.md) and [Level-Zero](extensions/Assert/level-zero.md)
+extensions.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+      // It's not defined if GPU thread stops execution
+      // Other GPU threads are left untouched
+   2. Specialized version of `__devicelib_assert_fail` is called
+   3. Device immediately signals to host (Low-Level Runtime)
+ - Host side:
+   1. The assert failure gets detected by Low-Level Runtime
+   2. Low-Level Runtime prints assert failure message to `stderr`
+   3. Low-Level Runtime calls `abort()`
+
+
+### Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a buffer
+based approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a dedicated assert buffer is allocated and implicit kernel
+argument is introduced. The argument is an accessor with `discard_read_write`
+or `discard_write` access mode. Accessor metadata is stored to program scope
```

Comment:
How will this work considering that program-scope variables must be const or constexpr?  From the spec:
> Variables with static storage duration that are odr-used inside a device function, must be const or
> constexpr and zero-initialized or constant-initialized.

Isn't some of the accessor metadata not known until run time?

Reply:
If I understand correctly, this limitation about `const` and `constexpr` was introduced because otherwise everyone can expect different results, like whether this non-const variable is shared between all work-items or not, whether changes to it on device side will be automatically propagated back to host, etc.

However, the idea here is to generate that global by the compiler - in this case we can overcome the limitation and we likely know what we are doing, so it should be safe

Reply:
How will this information be conveyed to the device?  Will there by a global variable in the SPIR-V that the kernel writes to when it is first called?